### PR TITLE
Change replacement of '--' to regex, to ignore autocorrect for <!--

### PR DESCRIPTION
**Pull request by justjack2347**
## Description

This changes the '--' replacement to a regex expression which does not allow '--' to autocorrect if the characters '<!' are in front of them. This should stop any problems with the term \
Changed the replacement string for '--' to a regex expression '/(?<!<!)--/'.

## Additional information

This is in regards to the issue #4340. Most changes recommended have been committed by @nathanlesage however I believe that this problem still persists. 
It is worth noting that the regex is potentially more confusing in the autocorrect tab of the preferences.

Additional contributors:
@josh-133 
@mr-verb

Tested on: Windows 11

### CODE REVIEW

**Comment by nathanlesage** [comment]
As much as I like the idea, this won't work, unfortunately.

* The autocorrect values must be simple strings. This RegExp is not going to work, it will simply look for the literal string of characters you put in there.
* Furthermore, allowing regular expressions would open the door to potential DoS attacks and unnecessarily complicate the logic.
* The correct way is to look for "protected nodes", as you will see in the corresponding file that contains the plugin's code here: https://github.com/Zettlr/Zettlr/blob/develop/source/common/modules/markdown-editor/commands/autocorrect.ts
* Lastly, as you will see there, I already added the last missing node that should not be autocorrected, inline `Comment`s. I did that after I closed issue #4340, when I noticed that I forgot to make that change.
* This being said, the bug fortunately does not happen anymore.

If you would like to improve the autocorrect, one thing that I would love to check out but don't have the time is if we still need the logic to check for YAML frontmatters inside the autocorrect function. Specifically, I am talking about this snippet:

https://github.com/Zettlr/Zettlr/blob/e920dee0741641c82707e04a30d9002ecb74d00a/source/common/modules/markdown-editor/commands/autocorrect.ts#L80-L84

I already added horizontal lines to the protected nodes, so I'm not even sure that this failsafe is necessary to detect horizontal rules. YAML frontmatters, on the other hand, are not yet part of the protectedNodes, but in general this approach should also work.

In other words, what you could do is try to get rid of the lines of code that I presented above, and instead add the correct nodes to the protected-check in the same file here:

https://github.com/Zettlr/Zettlr/blob/e920dee0741641c82707e04a30d9002ecb74d00a/source/common/modules/markdown-editor/commands/autocorrect.ts#L36-L45

I'm not 100% sure this would work, but someone needs to test this. If this works, this would clean up the function a lot. Additionally, this change would ensure that there are no autocorrections within the YAML frontmatter, which currently can happen (and therefore interfere with them).

**Comment by justjack2347** [comment]
Thanks @nathanlesage, I apologise for not realising that basic strings were the only form for replacements, for some reason I was under the belief I had tested its usage. And thanks for suggesting an issue to work on, we will have a look into it!

Using the current version, I still believe that I am experiencing the autocorrect bug. I think that the inline 'Comment' node is only being created after the final --> is created. This means that when using "ctrl+shift+c" to add a comment, there are no problems since it has finished creating the 'Comment' node which is protected. However, if you manually type a comment, the opening \, the inline 'Comment' is not generated and so no autocorrects are blocked.

I hope this makes sense and I am not missing something obvious. Cheers

**Comment by nathanlesage** [comment]
Ahh okay, yeah I see, the block comment exists as soon as you type the opening, and that was what I was thinking about and causing the confusion! Although in that case what people can do if they frequently do that is manually add a space in front of the `--` autocorrect to ensure this doesn't happen, so it's something the users can do if they do fancy! Nothing we should worry about.

And no need to apologize, I am more than happy about any help I can get with the app! :)
